### PR TITLE
Fix slow Trainer path with 4D attention mask

### LINKED ISSUE #32101: Using Trainer + a pretrained tokenizer + 4D attention mask is extremely slow
### System Info

transformers                  4.41.0

### Who can help?

@ArthurZucker

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

```
from transformers import LlamaForCausalLM, LlamaConfig, TrainingArguments, Trainer, AutoTokenizer
from datasets import IterableDataset
import numpy as np

model_config = LlamaConfig(
    vocab_size=10,
    hidden_size=384,
    num_hidden_layers=6,
    num_attention_heads=6,
    intermediate_size=1024,
    max_position_embeddings=1024,
)
model = LlamaForCausalLM(model_config)
tokenizer = AutoTokenizer.from_pretrained('facebook/opt-125m')

def get_data1():
    for i in range(10000):
        yield {'input_ids': np.zeros(1024, dtype=int), 'labels': np.zeros(1024, dtype=int), 'attention_mask': np.zeros((1, 1024, 1024), dtype=float)}

def get_data2():
    for i in range(10000):
        yield {'input_ids': np.zeros(1024, dtype=int), 'labels': np.zeros(1024, dtype=int), 'attention_mask': np.zeros((1024), dtype=int)}
    
ds_slow = IterableDataset.from_generator(get_data1).with_format('torch')
ds_fast = IterableDataset.from_generator(get_data2).with_format('torch')

training_args = TrainingArguments(max_steps=1, output_dir='./out', report_to=None, per_device_train_batch_size=32, gradient_accumulation_steps=32)
trainer1 = Trainer(model, training_args, train_dataset=ds_slow, tokenizer=tokenizer)
trainer2 = Trainer(model, training_args, train_dataset=ds_fast, tokenizer=tokenizer)

import cProfile
cProfile.run('trainer1.train()', './test_slow.profile')
cProfile.run('trainer2.train()', './test_fast.profile')
```

```
import pstats

# compare the two profiles
p1 = pstats.Stats('./test_slow.profile')
p2 = pstats.Stats('./test_fast.profile')
p1.sort_stats('cumtime').print_stats()
```
```
         1582200 function calls (1401111 primitive calls) in 340.112 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000  340.112  340.112 {built-in method builtins.exec}
        1    0.000    0.000  340.112  340.112 <string>:1(<module>)
        1    0.000    0.000  340.112  340.112 trainer.py:1784(train)
        1    0.017    0.017  340.112  340.112 trainer.py:1892(_inner_training_loop)
       33    0.001    0.000  326.171    9.884 data_loader.py:663(__iter__)
       33    0.001    0.000  325.473    9.863 data_loader.py:618(_fetch_batches)
 2486/265    0.001    0.000  325.428    1.228 {built-in method builtins.next}
       33    0.001    0.000  325.088    9.851 dataloader.py:625(__next__)
       33    0.725    0.022  325.083    9.851 dataloader.py:672(_next_data)
       33    0.002    0.000  323.988    9.818 fetch.py:24(fetch)
       33    0.000    0.000  320.979    9.727 trainer_utils.py:807(__call__)
       33    0.000    0.000  320.971    9.726 data_collator.py:270(__call__)
       33   16.982    0.515  320.971    9.726 data_collator.py:52(pad_without_fast_tokenizer_warning)
       33    0.005    0.000  303.989    9.212 tokenization_utils_base.py:3209(pad)
     6493  235.747    0.036  235.747    0.036 {built-in method torch.tensor}
      197    0.001    0.000  234.735    1.192 tokenization_utils_base.py:204(__init__)
      197    0.001    0.000  234.732    1.192 tokenization_utils_base.py:681(convert_to_tensors)
       99    0.000    0.000  234.730    2.371 tokenization_utils_base.py:718(as_tensor)
```

```
p2.sort_stats('cumtime').print_stats()
```

```
1567440 function calls (1386340 primitive calls) in 16.431 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000   16.431   16.431 {built-in method builtins.exec}
        1    0.000    0.000   16.431   16.431 <string>:1(<module>)
        1    0.000    0.000   16.431   16.431 [/opt/conda/lib/python3.11/site-packages/transformers/trainer.py:1784](https://vscode-remote+dev-002dcontainer-002b7b22766f6c756d654e616d65223a22616e61636f6e6461222c22666f6c646572223a22616e61636f6e6461227d-0040ssh-002dremote-002bida-002eece-002ewisc-002eedu.vscode-resource.vscode-cdn.net/opt/conda/lib/python3.11/site-packages/transformers/trainer.py:1784)(train)
        1    0.018    0.018   16.431   16.431 [/opt/conda/lib/python3.11/site-packages/transformers/trainer.py:1892](https://vscode-remote+dev-002dcontainer-002b7b22766f6c756d654e616d65223a22616e61636f6e6461222c22666f6c646572223a22616e61636f6e6461227d-0040ssh-002dremote-002bida-002eece-002ewisc-002eedu.vscode-resource.vscode-cdn.net/opt/conda/lib/python3.11/site-packages/transformers/trainer.py:1892)(_inner_training_loop)
       32    0.003    0.000   14.327    0.448 [/opt/conda/lib/python3.11/site-packages/transformers/trainer.py:3212](https://vscode-remote+dev-002dcontainer-002b7b22766f6c756d654e616d65223a22616e61636f6e6461222c22666f6c646572223a22616e61636f6e6461227d-0040ssh-002dremote-002bida-002eece-002ewisc-002eedu.vscode-resource.vscode-cdn.net/opt/conda/lib/python3.11/site-packages/transformers/trainer.py:3212)(training_step)
       32    0.001    0.000    8.830    0.276 [/opt/conda/lib/python3.11/site-packages/accelerate/accelerator.py:2093](https://vscode-remote+dev-002dcontainer-002b7b22766f6c756d654e616d65223a22616e61636f6e6461222c22666f6c646572223a22616e61636f6e6461227d-0040ssh-002dremote-002bida-002eece-002ewisc-002eedu.vscode-resource.vscode-cdn.net/opt/conda/lib/python3.11/site-packages/accelerate/accelerator.py:2093)(backward)
       32    0.000    0.000    8.829    0.276 [/opt/conda/lib/python3.11/site-packages/torch/_tensor.py:433](https://vscode-remote+dev-002dcontainer-002b7b22766f6c756d654e616d65223a22616e61636f6e6461222c22666f6c646572223a22616e61636f6e6461227d-0040ssh-002dremote-002bida-002eece-002ewisc-002eedu.vscode-resource.vscode-cdn.net/opt/conda/lib/python3.11/site-packages/torch/_tensor.py:433)(backward)
       32    0.000    0.000    8.829    0.276 [/opt/conda/lib/python3.11/site-packages/torch/autograd/__init__.py:149](https://vscode-remote+dev-002dcontainer-002b7b22766f6c756d654e616d65223a22616e61636f6e6461222c22666f6c646572223a22616e61636f6e6461227d-0040ssh-002dremote-002bida-002eece-002ewisc-002eedu.vscode-resource.vscode-cdn.net/opt/conda/lib/python3.11/site-packages/torch/autograd/__init__.py:149)(backward)
       32    8.827    0.276    8.827    0.276 {method 'run_backward' of 'torch._C._EngineBase' objects}
       33    0.000    0.000    4.546    0.138 [/opt/conda/lib/python3.11/site-packages/torch/cuda/memory.py:147](https://vscode-remote+dev-002dcontainer-002b7b22766f6c756d654e616d65223a22616e61636f6e6461222c22666f6c646572223a22616e61636f6e6461227d-0040ssh-002dremote-002bida-002eece-002ewisc-002eedu.vscode-resource.vscode-cdn.net/opt/conda/lib/python3.11/site-packages/torch/cuda/memory.py:147)(empty_cache)
       33    4.546    0.138    4.546    0.138 {built-in method torch._C._cuda_emptyCache}
 2486/265    0.001    0.000    1.469    0.006 {built-in method builtins.next}
       33    0.001    0.000    1.160    0.035 [/opt/conda/lib/python3.11/site-packages/accelerate/data_loader.py:663](https://vscode-remote+dev-002dcontainer-002b7b22766f6c756d654e616d65223a22616e61636f6e6461222c22666f6c646572223a22616e61636f6e6461227d-0040ssh-002dremote-002bida-002eece-002ewisc-002eedu.vscode-resource.vscode-cdn.net/opt/conda/lib/python3.11/site-packages/accelerate/data_loader.py:663)(__iter__)
       33    0.000    0.000    1.145    0.035 [/opt/conda/lib/python3.11/site-packages/accelerate/data_loader.py:618](https://vscode-remote+dev-002dcontainer-002b7b22766f6c756d654e616d65223a22616e61636f6e6461222c22666f6c646572223a22616e61636f6e6461227d-0040ssh-002dremote-002bida-002eece-002ewisc-002eedu.vscode-resource.vscode-cdn.net/opt/conda/lib/python3.11/site-packages/accelerate/data_loader.py:618)(_fetch_batches)
       33    0.000    0.000    1.136    0.034 [/opt/conda/lib/python3.11/site-packages/torch/utils/data/dataloader.py:625](https://vscode-remote+dev-002dcontainer-002b7b22766f6c756d654e616d65223a22616e61636f6e6461222c22666f6c646572223a22616e61636f6e6461227d-0040ssh-002dremote-002bida-002eece-002ewisc-002eedu.vscode-resource.vscode-cdn.net/opt/conda/lib/python3.11/site-packages/torch/utils/data/dataloader.py:625)(__next__)
       33    0.003    0.000    1.134    0.034 [/opt/conda/lib/python3.11/site-packages/torch/utils/data/dataloader.py:672](https://vscode-remote+dev-002dcontainer-002b7b22766f6c756d654e616d65223a22616e61636f6e6461222c22666f6c646572223a22616e61636f6e6461227d-0040ssh-002dremote-002bida-002eece-002ewisc-002eedu.vscode-resource.vscode-cdn.net/opt/conda/lib/python3.11/site-packages/torch/utils/data/dataloader.py:672)(_next_data)
       33    0.002    0.000    1.124    0.034 [/opt/conda/lib/python3.11/site-packages/torch/utils/data/_utils/fetch.py:24](https://vscode-remote+dev-002dcontainer-002b7b22766f6c756d654e616d65223a22616e61636f6e6461222c22666f6c646572223a22616e61636f6e6461227d-0040ssh-002dremote-002bida-002eece-002ewisc-002eedu.vscode-resource.vscode-cdn.net/opt/conda/lib/python3.11/site-packages/torch/utils/data/_utils/fetch.py:24)(fetch)
       32    0.000    0.000    0.955    0.030 [/opt/conda/lib/python3.11/site-packages/transformers/trainer.py:3254](https://vscode-remote+dev-002dcontainer-002b7b22766f6c756d654e616d65223a22616e61636f6e6461222c22666f6c646572223a22616e61636f6e6461227d-0040ssh-002dremote-002bida-002eece-002ewisc-002eedu.vscode-resource.vscode-cdn.net/opt/conda/lib/python3.11/site-packages/transformers/trainer.py:3254)(compute_loss)
...
```

### Expected behavior

Since the trace of the profiler is really long I only included the first few lines. 
I am running a small llama model on some dummy data, the only difference between the two datasets is that the slow version outputs 4D attention masks, which is a feature recently added in #27539. I am running both trainers for 1 iteration.

As you can see the slow run is 340s while the fast one runs in 16s. 

The slow version of the trainer is many times slower than the fast version. The problem probably lies in the default collator `DataCollatorWithPadding` (when there is a pretrained tokenizer), which calls `tokenizer.pad` on the 4D attention masks. When you takeaway either 1) the pretrained tokenizer or 2) the 4D attention mask, trainer runs much faster.

### PR DESCRIPTION
**Pull request by qflen**
Fixes #32101

`Trainer` with a pretrained tokenizer and a 4D `attention_mask` is dramatically slower than the 2D-mask path because `PreTrainedTokenizerBase.pad` routes the multi-dim per-sample mask through `to_py_obj` and then rebuilds it with `torch.tensor(deeply_nested_list)` in `BatchEncoding.convert_to_tensors`. That round trip is the dominant cost on the 4D path, and `_pad` itself cannot extend rank > 1 per-sample masks, so the conversion is wasted work.

This PR pops a multi-dim per-sample `attention_mask` out before the per-sample padding loop and stacks it back with `torch.stack` / `np.stack` at the end.

End-to-end on the OP's repro (CPU, `batch=32`, `grad_accum=32`, `seq=1024`, 1 optimizer step) the 4D `Trainer` step drops from 225.59 s to 114.62 s, within 13% of the 2D path's 101.52 s; the collator alone goes from ~3.65 s / batch to ~11.8 ms / batch.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
      Issue: https://github.com/huggingface/transformers/issues/32101.
      Maintainer invitation to open a PR: https://github.com/huggingface/transformers/issues/32101#issuecomment-4406013253.
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## Who can review?

@SunMarc (trainer) and @ArthurZucker (tokenizers, text models).

---

## Numbers

End-to-end Trainer benchmark (CPU, single optimizer step, `batch=32`, `grad_accum=32`, `seq=1024`) matching the OP's repro (tiny `LlamaForCausalLM`, 6 layers, hidden=384, vocab=10):

| `attention_mask` | before | after |
| --- | ---: | ---: |
| 2D `(seq,)` per sample | 99.53 s | 101.52 s |
| 4D `(1, q, q)` per sample | 225.59 s | 114.62 s |

Before the fix the 4D step was 2.27x slower than the 2D step. After the fix it is within 13%, since the remaining cost is dominated by the model forward / backward (~100 s on this CPU). On hardware where the model is faster (e.g. the OP's, where the 2D step took 16 s and the 4D step took 340 s), the fix recovers a much larger relative speedup at the same parity point, for the same reason: the collator stops being the bottleneck.

### Collator-only decomposition

`DataCollatorWithPadding` timed in isolation across three shapes; mean ms / batch over `n_repeat` calls after a warm-up:

| (batch, seq) | n_repeat | mask | before (ms / batch) | after (ms / batch) | speedup |
| --- | ---: | --- | ---: | ---: | ---: |
| (8, 128) | 200 | 2D | 0.32 | 0.33 | ~1.0x |
| (8, 128) | 200 | 4D `(1, q, q)` | 13.16 | 0.26 | **~50x** |
| (16, 256) | 50 | 2D | 1.20 | 1.16 | ~1.0x |
| (16, 256) | 50 | 4D `(1, q, q)` | 110.13 | 0.92 | **~120x** |
| (32, 1024) | 20 | 2D | 8.96 | 8.89 | ~1.0x |
| (32, 1024) | 20 | 4D `(1, q, q)` (OP shape) | 3650.73 | 11.80 | **~310x** |

The 2D path is unchanged within noise. The 4D speedup grows with shape because the dominant pre-fix cost is `torch.tensor` on a deeply nested Python list of size O(batch * q * kv); after the fix, a single `torch.stack` scales linearly with elements rather than super-linearly with Python overhead.

Numerical equivalence is verified by the new `test_4d_attention_mask_preserved` regression test: per-sample shape, dtype, and contents match the stacked batch slice exactly, for both torch and numpy inputs.

### CODE REVIEW

File: src/transformers/tokenization_utils_base.py [hunk 1]
```diff
@@ -2652,6 +2652,17 @@ def pad(
             # Call .keys() explicitly for compatibility with TensorDict and other Mapping subclasses
             encoded_inputs = {key: [example[key] for example in encoded_inputs] for key in encoded_inputs[0].keys()}
 
+        # `_pad` assumes 1D-per-sample inputs. A higher-rank per-sample `attention_mask`
+        # (e.g. a 4D causal mask of shape (batch, 1, q_len, kv_len)) is left untouched
+        # by the padding logic, but routing it through `to_py_obj` + `torch.tensor` on
+        # a deeply nested Python list dominates the runtime. Set it aside and stack it
+        # directly at the end of the batched path.
```
Comment (Rocketknight1) [inline src/transformers/tokenization_utils_base.py]:
```suggestion
        # Pop 4D nested-list attention masks and stack 
        # them at the end to avoid slow `to_py_obj`
```

**Review by Rocketknight1 — APPROVED** [review]
Looks good! Gonna shorten a classic Claude Comment but I'm happy with the new data collator logic otherwise.

**Comment by Rocketknight1** [comment]
Thanks for catching my style issue, merging!
